url9d1719148e_mrqsdavi_pegmachine_tgz-pJ8-peg_RegexJ8/src/peg/Regex.java

```diff
@@ -0,0 +1,616 @@
+	HashMap<String, Instrucao> instrucaoGramaticaTemporaria;
+	
+	public boolean useHeadFailOptimization = false;
+	public boolean usePartialCommitOptimization = false;
+	public boolean useSpanOptimization = false;
+	public boolean useFailTwiceOptimization = false;
+
+	@SuppressWarnings("static-access")
+	public Padrao compile(String padrao){
+		
+		InputStream is = new ByteArrayInputStream(padrao.getBytes());
+		AnalizadorLexico lexico = new AnalizadorLexico(is);
+		parser parser = new parser(lexico);
+		try {
+			parser.parse();
+		} catch (Exception e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		
+		return parser.padraoFinal;		
+	}
+	
+	public Integer find(Padrao padrao, String entrada){
+		Integer[] procura = findC(padrao, entrada);
+		if(procura != null){
+			return procura[0];
+		}
+		return null;
+	}
+	
+	public Integer find(String padrao, String entrada){
+		Integer[] procura = findC(padrao, entrada);
+		if(procura != null){
+			return procura[0];
+		}
+		return null;
+	}
+	
+	public Integer[] findC(Padrao padrao, String entrada){
+		Padrao padraoBusca = compile("S<- {"+padrao+"} / . S");
+		ArrayList<Instrucao> instrucoes = instrucoes(padraoBusca);
+		Maquina maquina = new Maquina(entrada, instrucoes);
+		maquina.run();
+		if(maquina.getCapturas().size() > 0){
+			return new Integer[]{maquina.getCapturas().get(0).getPosicaoInicial(), maquina.getCapturas().get(0).getPosicaoFinal()};
+		}
+		return null;
+	}
+	
+	public Integer[] findC(String padrao, String entrada){
+		Padrao padraoCompilado = compile(padrao);
+		return findC(padraoCompilado, entrada);
+	}
+	
+	public ArrayList<Instrucao> instrucoes(Padrao padrao){
+
+		instrucaoGramatica = new HashMap<>();
+		instrucaoGramaticaTemporaria = new HashMap<>();
+		
+		
+		if(padrao.getTipo()==TipoPadrao.GRAMATICA){
+			instrucaoGramaticaTemporaria.put(padrao.gramatica().getNome(), new ICall("TEMP"));
+			for(int i = 0; i < padrao.gramatica().getSubgramaticas().size(); i++){
+				Gramatica subgramatica = padrao.gramatica().getSubgramaticas().get(i);
+				instrucaoGramaticaTemporaria.put(subgramatica.getNome(), new ICall("TEMP"));
+			}
+		}
+
+		
+		end = new IEnd();
+		next = end;
+		ArrayList<Instrucao> retorno = instrucoesDoPadrao(padrao);
+		retorno.add(end);
+
+		for(int i = 0; i < retorno.size(); i++){	
+			
+			Instrucao instrucaoAtual = retorno.get(i);
+			
+			for(Entry<String, Instrucao> entry : instrucaoGramaticaTemporaria.entrySet()) {
+			    String key = entry.getKey();
+			    Instrucao value = entry.getValue();
+			    
+			    if(instrucaoAtual.getInstrucaoDesvio() == value){
+			    	instrucaoAtual.setInstrucaoDesvio(instrucaoGramatica.get(key));
+			    }
+			    
+			}
+			
+			if(instrucaoAtual.getInstrucaoDesvio()!=null){
+				instrucaoAtual.setIndexDesvio(retorno.indexOf(instrucaoAtual.getInstrucaoDesvio()));
+			}
+		}
+		
+		return retorno;
+		
+	}
+	
+	public ArrayList<Instrucao> instrucoesDoPadrao(Padrao padrao){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 149 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.